unison-src/transcripts/command-replace.output.md

```diff
@@ -55,9 +55,29 @@ Test that replace works with types
 
   Done.
 
+.scratch> find
+
+  1. type X
+  2. X.One : Nat -> Nat -> X
+  3. type Y
+  4. Y.Two : Nat -> Nat -> X
+  5. x : Nat
+  6. y : Nat
+  
+
+.scratch> view.patch patch
+
+  Edited Types: X#d97e0jhkmd -> X
+  
+  Edited Terms: #jk19sm5bf8 -> x
+  
+  Tip: To remove entries from a patch, use
+       delete.term-replacement or delete.type-replacement, as
+       appropriate.
+
 .scratch> view X
 
-  type X = Two Nat Nat
+  type X = One Nat Nat
```

Comment:
It took me a while staring at this to understand that this is the correct output. I think this transcript example is kind of weird, in that it is replacing one type with another, but both types are under the current namespace, so they both get replaced with the target type, but with two different names for the constructors.

Reply:
TL;DR: the old output was wrong / a consequence of constructor propagation not working. The new output is correct, but also, the transcript is kind of a weird example.

Reply:
Doesn't it seem wrong or at least super confusing that the original constructor name (which doesn't appear in the replacement type) is reused, but given a new type? I'm imagining things like `None : a -> Maybe a`.  

Is this something we should track separately?

Reply:
I think it could be confusing but given that patches are nameless mappings it's the correct / only behavior. When you say `replace Type1 Type2`, you're mapping one hash to another (and by extension, to the constructors of that hash). Wherever the patch is applied, if names exist for the old hash it moves those names to the new hash. So this is just the same thing. Term replacement works the same way, incidentally.

So it is a feature of how patches are "nameless" and can be applied even to a namespace where the names are different than the namespace where the patch was prepared, and doesn't mess with the local names chosen within that namespace for the same definitions.

I do think sometimes you do want to update a thing, and also rename it (or perhaps suggest a renaming), but right now that's not something patches talk about directly. It could be interesting to make patches capable of doing more stuff, or I think a codebase ability in Unison would do it, so you could express more interesting codebase transformations.

Reply:
I get it for replacing types and I get  for replacing terms (though, maybe we should rethink this), but I don’t get it for replacing constructors; it still seems super wrong / random to me.  What if there’s more than one constructor? Is there any way to predict which old constructor name will correspond to which new constructor at the end?

Reply:
> What if there’s more than one constructor? Is there any way to predict which old constructor name will correspond to which new constructor at the end?

Yes, it's just whatever the constructor mapping is that is established before propagation starts, or (when patches support this) whatever constructor mapping is in the patch. In this PR (and eventually the `update` command, when patches support this directly) the identity of the same named constructor is mapped to the identity of the new same named constructor, and then this nameless mapping can be applied to other namespaces, preserving whatever names exist for those constructors in those namespaces.

An example I was thinking about is like - you have two namespaces, one with the identifiers all in English, and another where the identifiers are in Spanish.

You can create a patch in the English namespace, then apply it to the Spanish namespace, mapping constructors that have different names but "the same role" in their respective aliases of the same type definition. So, it's not a random mapping that is like, positional or something.